### PR TITLE
Add generic packages

### PR DESCRIPTION
**Pull request by dimkouv**
Added generic:

- Set
- Slice Methods: partition, contains, value_at, min
- OrderedSet

### CODE REVIEW

File: set/set.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+			return nil, err
+		}
+
+		sl = append(sl, n)
+	}
+
+	return sl, nil
+}
+
+func (s *Set[T]) InitFromSlice(values []T) *Set[T] {
```
Comment (covain) [inline set/set.go:102]: What is difference between this func and InitFromValues?

File: set/set.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+func (s *Set[T]) ToSlice() []T {
+	sl := make([]T, 0, len(s.values))
+
+	for v := range s.values {
+		sl = append(sl, v)
+	}
+	return sl
+}
+
+func (s *Set[T]) ToIntSlice() ([]int, error) {
```
Comment (covain) [inline set/set.go:87]: No need to have this func, it duplicates ToSlice()

File: slice/partition.go [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+		return [][]T{}
+	}
+
+	partitions := make([][]T, 0, len(s)/partitionSize)
+
+	for partitionNumber := 0; ; partitionNumber++ {
+		left := partitionNumber * partitionSize
+		if left >= len(s) {
+			break
+		}
```
Comment (covain) [inline slice/partition.go:15]: Hard to understand. Maybe it makes sense to use for without conditions?
```
partitionNumber := 0
for {
    left := partitionNumber * partitionSize
		if left >= len(s) {
			break
		}
...
partitionNumber++
}
```

File: slice/search.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		if v == val {
+			return true
+		}
+	}
+
+	return false
+}
+
+// ValueAt returns if exists values[idx] else the fallback value.
+func ValueAt[T any](idx int, values []T, fallback T) T {
```
Comment (covain) [inline slice/search.go:17]:
```suggestion
func ValueAt[T any](idx int, values []T, default T) T {
```
?

File: slice/unique.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package slice
+
+type Unique[T comparable] struct {
```
Comment (covain) [inline slice/unique.go:3]: why may we need it?

**Review by covain — CHANGES_REQUESTED** [review]
Left comments that require code adjustments

File: slice/search.go (hunk 4 above)
Comment (covain) [reply slice/search.go:17]: then either `fallbackValue` or `defaultValue`

File: set/set.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	sl := make([]T, 0, len(s.values))
+
+	for v := range s.values {
+		sl = append(sl, v)
+	}
+	return sl
+}
+
+// ResetFromSlice will clear all the set values and push all the provided values.
+func (s *Set[T]) ResetFromSlice(values []T) *Set[T] {
```
Comment (covain) [inline set/set.go:86]:
```
values1 := []int{1, 2}
values2 := []int{3,4}

s := [int]NewFromValues(values1...)
// so the difference is:
s = [int]NewFromValues(values2...)
// or
s = s.ResetFromSlice(values2)
```
Doesn't seem to me as we need this function

File: slice/partition.go [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+	for {
+		left := len(partitions) * partitionSize
+		if left >= len(s) {
+			break
+		}
+
+		right := left + partitionSize
+		if right > len(s) {
+			right = len(s)
+		}
```
Comment (covain) [inline slice/partition.go:20]:
```suggestion
		right := math.Min(left + partitionSize, len(s))
```

File: slice/unique.go (hunk 5 above)
Comment (covain) [reply slice/unique.go:3]: Let's rename it to OrderedSet

File: slice/partition.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
+package slice
+
+// Partition creates partitions of a standard maximum size.
+func Partition[T any](s []T, partitionSize int) [][]T {
+	if len(s) == 0 || partitionSize <= 0 {
+		return [][]T{}
+	}
+
+	partitions := make([][]T, 0, len(s)/partitionSize)
```
Comment (DeimonDB) [inline slice/partition.go:9]: the capacity should be rounded up `make([][]T, 0, (len(s) + partitionSize - 1)/partitionSize)`

File: slice/partition.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+			break
+		}
+
+		right := Min(left+partitionSize, len(s))
+
+		part := s[left:right]
+		partition := make([]T, len(part))
+		copy(partition, part)
+
+		if len(partition) == 0 {
```
Comment (DeimonDB) [inline slice/partition.go:23]: this `if` will never be executed right?

File: slice/partition.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+	partitions := make([][]T, 0, len(s)/partitionSize)
+
+	for {
+		left := len(partitions) * partitionSize
+		if left >= len(s) {
+			break
+		}
+
+		right := Min(left+partitionSize, len(s))
```
Comment (covain) [inline slice/partition.go:17]: :fire:

**Review by DeimonDB — CHANGES_REQUESTED** [review]

**Review by covain — DISMISSED** [review]
Awesome work!

**Review by DeimonDB — APPROVED** [review]
LGTM!

**Pull request merged**